In [1]:
# Enable code formatting using external plugin: nb_black.
%reload_ext nb_black

<IPython.core.display.Javascript object>

![Pandas Tutorial](pandas.png)

# Pandas Tutorial - PART 1

**Ref: [Pandas Tutorials][1] by [Corey Schafer](https://coreyms.com/)**

[1]: https://youtube.com/playlist?list=PL-osiE80TeTsWmV9i9c58mdDCSskIFdDS

## Introduction

[`pandas`](https://pandas.pydata.org/) is a software library written for the Python programming language for data manipulation and analysis. In particular, **it offers data structures and operations for manipulating numerical tables and time series**. The name is derived from the term "_panel data_", an econometrics term for data sets that include observations over multiple time periods for the same individuals.

Important data types used in `pandas` are:

1. Series
2. DataFrame

### Install & load Pandas library

Run below command if pandas is not already installed.
```cmd
pip install pandas
```
or
```cmd
pip3 install pandas
```

In [2]:
import pandas as pd

print("Pandas version:", pd.__version__)

Pandas version: 1.3.4


<IPython.core.display.Javascript object>

### Introduction to `Series` and `DataFrame` using _Stackoverflow_ data

By default printing a `DataFrame` displays:

1. Total 10 rows which contains top 5 and bottom 5 rows.
2. Total 20 columns which contains first 10 and last 10 columns.

In [3]:
# Load Stackoverflow data into a DataFrame from the csv file.
df = pd.read_csv("./data/Stackoverflow/data/survey_results_public.csv")
df

Respondent                                         MainBranch Hobbyist  \
0               1             I am a student who is learning to code      Yes   
1               2             I am a student who is learning to code       No   
2               3  I am not primarily a developer, but I write co...      Yes   
3               4                     I am a developer by profession       No   
4               5                     I am a developer by profession      Yes   
...           ...                                                ...      ...   
88878       88377                                                NaN      Yes   
88879       88601                                                NaN       No   
88880       88802                                                NaN       No   
88881       88816                                                NaN       No   
88882       88863                                                NaN      Yes   

                                             OpenSourcer  \
0                                                  Never   
1                                Less than once per year   
2                                                  Never   
3                                                  Never   
4                             Once a month or more often   
...                                                  ...   
88878  Less than once a month but more than once per ...   
88879                                              Never   
88880                                              Never   
88881                                              Never   
88882                            Less than once per year   

                                              OpenSource  \
0      The quality of OSS and closed source software ...   
1      The quality of OSS and closed source software ...   
2      The quality of OSS and closed source software ...   
3      The quality of OSS and closed source software ...   
4      OSS is, on average, of HIGHER quality than pro...   
...                                                  ...   
88878  The quality of OSS and closed source software ...   
88879  The quality of OSS and closed source software ...   
88880                                                NaN   
88881  OSS is, on average, of HIGHER quality than pro...   
88882  OSS is, on average, of HIGHER quality than pro...   

                                              Employment  \
0                 Not employed, and not looking for work   
1                     Not employed, but looking for work   
2                                     Employed full-time   
3                                     Employed full-time   
4                                     Employed full-time   
...                                                  ...   
88878             Not employed, and not looking for work   
88879                                                NaN   
88880                                 Employed full-time   
88881  Independent contractor, freelancer, or self-em...   
88882             Not employed, and not looking for work   

                      Country         Student  \
0              United Kingdom              No   
1      Bosnia and Herzegovina  Yes, full-time   
2                    Thailand              No   
3               United States              No   
4                     Ukraine              No   
...                       ...             ...   
88878                  Canada              No   
88879                     NaN             NaN   
88880                     NaN             NaN   
88881                     NaN             NaN   
88882                   Spain  Yes, full-time   

                                                 EdLevel  \
0                              Primary/elementary school   
1      Secondary school (e.g. American high school, G...   
2               Bachelor’s degree (BA, BS, B.Eng., etc.)   
3               Bachelor’s degree (BA, BS,

<IPython.core.display.Javascript object>

#### `shape` attribute of the `Dataframe`

In [4]:
df.shape

(88883, 85)

<IPython.core.display.Javascript object>

This `Dataframe` has 88883 rows and 85 columns.

#### Metadata of `Dataframe`

Use `info()` to display **total number of rows and columns with datatype**.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88883 entries, 0 to 88882
Data columns (total 85 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Respondent              88883 non-null  int64  
 1   MainBranch              88331 non-null  object 
 2   Hobbyist                88883 non-null  object 
 3   OpenSourcer             88883 non-null  object 
 4   OpenSource              86842 non-null  object 
 5   Employment              87181 non-null  object 
 6   Country                 88751 non-null  object 
 7   Student                 87014 non-null  object 
 8   EdLevel                 86390 non-null  object 
 9   UndergradMajor          75614 non-null  object 
 10  EduOther                84260 non-null  object 
 11  OrgSize                 71791 non-null  object 
 12  DevType                 81335 non-null  object 
 13  YearsCode               87938 non-null  object 
 14  Age1stCode              87634 non-null

<IPython.core.display.Javascript object>

> Note: Most of the time datatype `object` means `String`.

#### Fetch first five rows from `Dataframe`

In [6]:
df.head(6)  # Default value is 5 if 6 was not passed.

Respondent                                         MainBranch Hobbyist  \
0           1             I am a student who is learning to code      Yes   
1           2             I am a student who is learning to code       No   
2           3  I am not primarily a developer, but I write co...      Yes   
3           4                     I am a developer by profession       No   
4           5                     I am a developer by profession      Yes   
5           6  I am not primarily a developer, but I write co...      Yes   

                  OpenSourcer  \
0                       Never   
1     Less than once per year   
2                       Never   
3                       Never   
4  Once a month or more often   
5                       Never   

                                          OpenSource  \
0  The quality of OSS and closed source software ...   
1  The quality of OSS and closed source software ...   
2  The quality of OSS and closed source software ...   
3  The quality of OSS and closed source software ...   
4  OSS is, on average, of HIGHER quality than pro...   
5  The quality of OSS and closed source software ...   

                               Employment                 Country  \
0  Not employed, and not looking for work          United Kingdom   
1      Not employed, but looking for work  Bosnia and Herzegovina   
2                      Employed full-time                Thailand   
3                      Employed full-time           United States   
4                      Employed full-time                 Ukraine   
5                      Employed full-time                  Canada   

          Student                                            EdLevel  \
0              No                          Primary/elementary school   
1  Yes, full-time  Secondary school (e.g. American high school, G...   
2              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
3              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
4              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
5              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                      UndergradMajor  ...  \
0                                                NaN  ...   
1                                                NaN  ...   
2                      Web development or web design  ...   
3  Computer science, computer engineering, or sof...  ...   
4  Computer science, computer engineering, or sof...  ...   
5                          Mathematics or statistics  ...   

                             WelcomeChange  \
0  Just as welcome now as I felt last year   
1  Just as welcome now as I felt last year   
2  Just as welcome now as I felt last year   
3  Just as welcome now as I felt last year   
4  Just as welcome now as I felt last year   
5  Just as welcome now as I felt last year   

                                        SONewContent   Age Gender Trans  \
0  Tech articles written by other developers;Indu...  14.0    Man    No   
1  Tech articles written by other developers;Indu...  19.0    Man    No   
2  Tech meetups or events in your area;Courses on...  28.0    Man    No   
3  Tech articles written by other developers;Indu...  22.0    Man    No   
4  Tech meetups or events in your area;Courses on...  30.0    Man    No   
5  Tech articles written by other developers;Indu...  28.0    Man    No   

                 Sexuality                                 Ethnicity  \
0  Straight / Heterosexual                                       NaN   
1  Straight / Heterosexual                                       NaN   
2  Straight / Heterosexual                                       NaN   
3  Straight / Heterosexual              White or of European descent   
4  Straight / Heterosexual  White or of European descent;Multiracial   
5  Straight / Heterosexual                                East Asian   

  Dependents           SurveyLength                  SurveyEase  
0

<IPython.core.display.Javascript object>

#### Fetch last five rows from `Dataframe`

In [7]:
df.tail(3)  # Default value is 5 if 3 was not passed.

Respondent MainBranch Hobbyist              OpenSourcer  \
88880       88802        NaN       No                    Never   
88881       88816        NaN       No                    Never   
88882       88863        NaN      Yes  Less than once per year   

                                              OpenSource  \
88880                                                NaN   
88881  OSS is, on average, of HIGHER quality than pro...   
88882  OSS is, on average, of HIGHER quality than pro...   

                                              Employment Country  \
88880                                 Employed full-time     NaN   
88881  Independent contractor, freelancer, or self-em...     NaN   
88882             Not employed, and not looking for work   Spain   

              Student                             EdLevel  \
88880             NaN                                 NaN   
88881             NaN                                 NaN   
88882  Yes, full-time  Professional degree (JD, MD, etc.)   

                                          UndergradMajor  ...  \
88880                                                NaN  ...   
88881                                                NaN  ...   
88882  Computer science, computer engineering, or sof...  ...   

                                  WelcomeChange  \
88880                                       NaN   
88881                                       NaN   
88882  Somewhat less welcome now than last year   

                                            SONewContent   Age Gender Trans  \
88880                                                NaN   NaN    NaN   NaN   
88881                                                NaN   NaN    NaN   NaN   
88882  Tech articles written by other developers;Indu...  18.0    Man    No   

                     Sexuality  \
88880                      NaN   
88881                      NaN   
88882  Straight / Heterosexual   

                                               Ethnicity Dependents  \
88880                                                NaN        NaN   
88881                                                NaN        NaN   
88882  Hispanic or Latino/Latina;White or of European...         No   

                SurveyLength SurveyEase  
88880                    NaN        NaN  
88881                    NaN        NaN  
88882  Appropriate in length       Easy  

[3 rows x 85 columns]

<IPython.core.display.Javascript object>

### Options

Some important options in `pandas` to alter the output format:

1. display.max_columns
2. display.max_rows
3. display.width

Syntax:

```python
pd.options.display.max_columns = 20
pd.options.display.max_rows = 10
pd.options.display.width = 130
```

#### Modify option `display.max_columns` to display required number of columns

In [8]:
old_max_cols = pd.options.display.max_columns
print("Old value in option `display.max_columns`:", old_max_cols)

(row_count, col_count) = df.shape
pd.options.display.max_columns = col_count
print("New value in option `display.max_columns`:", col_count)

Old value in option `display.max_columns`: 20
New value in option `display.max_columns`: 85


<IPython.core.display.Javascript object>

Disply first row to see all the columns.

In [9]:
df.head(1)

Respondent                              MainBranch Hobbyist OpenSourcer  \
0           1  I am a student who is learning to code      Yes       Never   

                                          OpenSource  \
0  The quality of OSS and closed source software ...   

                               Employment         Country Student  \
0  Not employed, and not looking for work  United Kingdom      No   

                     EdLevel UndergradMajor  \
0  Primary/elementary school            NaN   

                                            EduOther OrgSize DevType  \
0  Taught yourself a new language, framework, or ...     NaN     NaN   

  YearsCode Age1stCode YearsCodePro CareerSat JobSat MgrIdiot MgrMoney  \
0         4         10          NaN       NaN    NaN      NaN      NaN   

  MgrWant JobSeek LastHireDate LastInt FizzBuzz JobFactors ResumeUpdate  \
0     NaN     NaN          NaN     NaN      NaN        NaN          NaN   

  CurrencySymbol CurrencyDesc  CompTotal CompFreq  ConvertedComp  WorkWeekHrs  \
0            NaN          NaN        NaN      NaN            NaN          NaN   

  WorkPlan WorkChallenge WorkRemote WorkLoc ImpSyn CodeRev  CodeRevHrs  \
0      NaN           NaN        NaN     NaN    NaN     NaN         NaN   

  UnitTests PurchaseHow PurchaseWhat               LanguageWorkedWith  \
0       NaN         NaN          NaN  HTML/CSS;Java;JavaScript;Python   

                            LanguageDesireNextYear DatabaseWorkedWith  \
0  C;C++;C#;Go;HTML/CSS;Java;JavaScript;Python;SQL             SQLite   

  DatabaseDesireNextYear PlatformWorkedWith   PlatformDesireNextYear  \
0                  MySQL      MacOS;Windows  Android;Arduino;Windows   

  WebFrameWorkedWith WebFrameDesireNextYear MiscTechWorkedWith  \
0       Django;Flask           Flask;jQuery            Node.js   

  MiscTechDesireNextYear                  DevEnviron    OpSys  \
0                Node.js  IntelliJ;Notepad++;PyCharm  Windows   

                Containers BlockchainOrg BlockchainIs BetterLife  \
0  I do not use containers           NaN          NaN        Yes   

                                   ITperson OffOn SocialMedia Extraversion  \
0  Fortunately, someone else has that title   Yes     Twitter       Online   

  ScreenName SOVisit1st                      SOVisitFreq  \
0   Username       2017  A few times per month or weekly   

                                           SOVisitTo        SOFindAnswer  \
0  Find answers to specific questions;Learn how t...  3-5 times per week   

                      SOTimeSaved  SOHowMuchTime SOAccount SOPartFreq  \
0  Stack Overflow was much faster  31-60 minutes        No        NaN   

                                              SOJobs  \
0  No, I didn't know that Stack Overflow had a jo...   

                              EntTeams   SOComm  \
0  No, and I don't know what those are  Neutral   

                             WelcomeChange  \
0  Just as welcome now as I felt last year   

                                        SONewContent   Age Gender Trans  \
0  Tech articles written by other developers;Indu...  14.0    Man    No   

                 Sexuality Ethnicity Dependents           SurveyLength  \
0  Straight / Heterosexual       NaN         No  Appropriate in length   

                   SurveyEase  
0  Neither easy nor difficult

<IPython.core.display.Javascript object>

### Schema of the _Stackoverflow_ data

_survey_results_schema.csv_ file contains details on each minified-column names in the _survey_results_public.csv_ file, acting like a schema.

In [10]:
# Load schema from csv file.
sdf = pd.read_csv("./data/Stackoverflow/data/survey_results_schema.csv")

<IPython.core.display.Javascript object>

#### Modify option `display.max_rows` to display required number of rows

In [11]:
old_max_rows = pd.options.display.max_rows
print("Old value in option `display.max_rows`:", old_max_rows)

(row_count, col_count) = sdf.shape
pd.options.display.max_rows = row_count
print("New value in option `display.max_rows`:", row_count)

Old value in option `display.max_rows`: 60
New value in option `display.max_rows`: 85


<IPython.core.display.Javascript object>

#### Modify styling to left align column values

Applying style on a dataframe returns a [Styler object](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.io.formats.style.Styler.html#pandas.io.formats.style.Styler), **not a DataFrame**. You cannot apply further style operations on that.

Ref: [Stackoverflow answer.](https://stackoverflow.com/questions/67869034/styler-object-has-no-attribute-style#answer-67869252)

In [12]:
s_sdf = sdf.style.set_properties(**{"text-align": "left"})  # Styled Schema Dataframe
s_sdf

<IPython.core.display.Javascript object>